src/PartitionLoops.cpp

```diff
@@ -217,16 +217,50 @@ struct Simplification {
     Interval interval;
 };
 
+template<typename T>
+class ExprUsesValidBuffers : public IRGraphVisitor {
+    using IRGraphVisitor::visit;
+
+    const Scope<T> &allocated_buffers;
+
+    void visit(const Load *op) {
+        if (!allocated_buffers.contains(op->name)) {
+            valid = false;
+        } else {
+            IRGraphVisitor::visit(op);
+        }
+    }
+public:
+    ExprUsesValidBuffers(const Scope<T> &buffers) : allocated_buffers(buffers), valid(true) {}
+    bool valid;
+};
+
+/** Check if any references to buffers in an expression is valid (i.e. the buffers
+ * have already been allocated).
+ */
+template<typename T>
+bool expr_uses_valid_buffers(Expr e, const Scope<T> &buffers) {
```

Comment:
This should just take a Scope<int> and not be a template, same with the visitor above.